File: two_sum.cxx
#include <unordered_map>
#include <vector>

using namespace std;

class Solution
{
public:
  vector<int> twoSum(vector<int> &nums, int target)
  {
    unordered_map<int, int> nm;
    int n = nums.size();

    // Build the hash table
    for (int i = 0; i < n; i++)
    {
      nm[nums[i]] = i;
    }

    // Find the complement
    for (int i = 0; i < n; i++)
    {
      int complement = target - nums[i];
      if (nm.count(complement) && nm[complement] != i)
      {
        return {i, nm[complement]};
      }
    }

    return {};
  }
};

File: two_sum.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        
        HashMap<Integer, Integer> mp = new HashMap<>();

        // Iterate through input sequence
        for (int i = 0; i < nums.length; i++) {
            mp.put(nums[i], i);
        }

        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (mp.containsKey(complement) && mp.get(complement) != i) {
                return new int[]{i, mp.get(complement)};
            }
        }

        return new int[]{};
    
    }
}
